File: 2020/src/day25.cpp
#include "day25.h"

ResultBase* Day25::SolveFirst() const
{
	auto [cardKey, doorKey] = ParseInput();
	uint64_t cardLoop = CalcLoopSize(cardKey);
	return new Result(CalcEncryptionKey(doorKey, cardLoop));
}

ResultBase* Day25::SolveSecond() const
{
	//Day 25 only has 1 puzzle
	return nullptr;
}

uint64_t Day25::CalcLoopSize(uint64_t key) const
{
	uint64_t v = 1;
	for(uint64_t i = 0;; i++)
	{
		v = v * 7 % 20201227;
		if(v == key)
		{
			return i + 1;
		}
	}
	return 0;
}

uint64_t Day25::CalcEncryptionKey(uint64_t subjectNumber, uint64_t loopSize) const
{
	uint64_t v = 1;
	for(uint64_t i = 0; i < loopSize; i++)
	{
		v = v * subjectNumber % 20201227;
	}
	return v;
}

std::pair<uint64_t, uint64_t> Day25::ParseInput() const
{
	return std::make_pair(std::stoull(input.lines[0]), std::stoull(input.lines[1]));
}

File: 2024/src/day25.fs
module Day25
open Input
open Common

[<Literal>]
let name = "Code Chronicle"

type Schematic = Lock | Key

let parseInput (input: Input) = 
    input.lines
    |> List.split ((=) "")
    |> List.map (fun n -> 
        let lock = n |> List.head |> Seq.forall ((=) '#')
        let ref = if lock then '.' else '#'
        let depths = 
            n
            |> List.map List.ofSeq
            |> List.transpose
            |> List.map (List.filter ((=) ref) >> List.length)
        if lock then (Lock, depths) else (Key, depths)
    )

let matches (lock: int list) (key: int list) = 
    (lock, key)
    ||> List.zip
    |> List.forall (fun (lDepth, kDepth) -> lDepth >= kDepth)

let solveFirst (input: Input) = 
    let schematics = input |> parseInput
    let locks = schematics |> List.filter (fst >> (=) Lock) |> List.map snd
    let keys = schematics |> List.filter (fst >> (=) Key) |> List.map snd
    (locks, keys)
    ||> List.allPairs
    |> List.filter (fun (lock, key) -> matches lock key)
    |> List.length

let solveSecond (input: Input) = 
    //Day 25 only has 1 puzzle
    null :> obj
